src/main/java/net/neoforged/neoforge/common/extensions/IBlockGetterExtension.java

```diff
@@ -59,11 +59,12 @@ default AuxiliaryLightManager getAuxLightManager(ChunkPos pos) {
     }
 
     /**
-     * Retrieves the model data manager for this level.
-     * This will be {@code null} on a server level.
+     * Retrieves model data for a block at the given position.
+     *
+     * @param pos the position to query
+     * @return the model data at this position, or {@link ModelData#EMPTY} if none exists
      */
-    @Nullable
-    default ModelDataManager getModelDataManager() {
-        return null;
+    default ModelData getModelData(BlockPos pos) {
+        return ModelData.EMPTY;
```

Comment:
My proposal was based on Fabric API's design. There, the default implementation of this method works like `BlockEntity be = getBlockEntity(pos); return be != null : be.getModelData() : ModelData.EMPTY`. While this isn't an appropriate implementation for asynchronous scenarios and doesn't consider the current level's cache, it could still be useful as many modded scenarios I have seen are not asynchronous and are independent of the current level. Either default implementation is fine for my use cases but this is worth considering.

Reply:
This seems like a reasonable choice as well; I'll leave it open for others to comment as I personally am undecided which is better.

Reply:
I don't think that's a good idea. One of the most prevalent uses of custom `Block(AndTint)Getter`s is for tricks related to `IBakedModelExtension#getModelData()` and if a mod forgets to override the extension method in their custom implementation, then you've got an almost untraceable bug that reintroduces the exact concurrency issues the model data system aims to prevent. If it defaults to empty data though, it's almost immediately obvious what's wrong when a mod that heavily relies on model data interacts with one that uses a custom `Block(AndTint)Getter` (I had this case with FramedBlocks and Athena where the latter didn't forward the `ModelDataManager` from the wrapped `BlockGetter`).

Reply:
To add to what XFact said, there can also be mods that have a very expensive model data computation and rely on the cache in the level. It's not a good idea to bypass such a cache.